### PR TITLE
Bluetooth: audio: BAP: allow for multiple callbacks

### PR DESCRIPTION
**Pull request by kruithofa**
The existing implementation allows for only one callback per procedure. This creates problems when for example adding CAP functionality. The callbacks are here implemented as an SLIST

### CODE REVIEW

File: subsys/bluetooth/audio/bap_broadcast_assistant.c [hunk 1]
```diff
@@ -381,11 +383,18 @@ static uint8_t notify_handler(struct bt_conn *conn,
...
-		broadcast_assistant.past_avail[index] = false;
-		broadcast_assistant_cbs->recv_state_removed(conn, 0,
-							    broadcast_assistant.src_ids[index]);
+	} else {
+		struct bt_bap_broadcast_assistant_cb *listener, *next;
+
+		SYS_SLIST_FOR_EACH_CONTAINER_SAFE(&broadcast_assistant_cbs,
+						  listener, next, _node) {
+			if (listener->recv_state_removed) {
+				broadcast_assistant.past_avail[index] = false;
```
Comment (Thalley) [inline subsys/bluetooth/audio/bap_broadcast_assistant.c:392]: This looks like a bug: `broadcast_assistant.past_avail[index] = false;` is only done if there is a callback, but it should always be done, right? Suggest to fix in this PR, but as a separate commit

File: subsys/bluetooth/audio/bap_broadcast_assistant.c [hunk 2]
```diff
@@ -611,10 +640,11 @@ static uint8_t service_discover_func(struct bt_conn *conn,
...
 			broadcast_assistant.busy = false;
-
-			if (broadcast_assistant_cbs != NULL &&
-			    broadcast_assistant_cbs->discover != NULL) {
-				broadcast_assistant_cbs->discover(conn, err, 0);
+			SYS_SLIST_FOR_EACH_CONTAINER_SAFE(&broadcast_assistant_cbs,
+							  listener, next, _node) {
+				if (listener->discover) {
+					listener->discover(conn, err, 0);
+				}
```
Comment (Thalley) [inline subsys/bluetooth/audio/bap_broadcast_assistant.c:647]: Suggest to create helper functions for calling the callback, as we've done for e.g. VCP and MICP: 
```c
static void vcp_vol_ctlr_flags_changed(struct bt_vcp_vol_ctlr *vol_ctlr, int err)
{
	struct bt_vcp_vol_ctlr_cb *listener, *next;

	SYS_SLIST_FOR_EACH_CONTAINER_SAFE(&vcp_vol_ctlr_cbs, listener, next, _node) {
		if (listener->flags) {
			listener->flags(vol_ctlr, err, err == 0 ? vol_ctlr->flags : 0);
		}
	}
}
```

This applies for all callbacks

File: subsys/bluetooth/audio/bap_broadcast_assistant.c [hunk 3]
```diff
@@ -861,9 +892,40 @@ int bt_bap_broadcast_assistant_discover(struct bt_conn *conn)
...
+	}
+
+	SYS_SLIST_FOR_EACH_CONTAINER(&broadcast_assistant_cbs, tmp, _node) {
+		if (tmp == cb) {
+			LOG_DBG("Already registered");
+			return -EALREADY;
+		}
+	}
+
+	LOG_DBG("Adding callbacks to the list");
```
Comment (Thalley) [inline subsys/bluetooth/audio/bap_broadcast_assistant.c:911]:
```suggestion
```

**Review by Thalley — CHANGES_REQUESTED** [review]
Some suggestions, but otherwise looks good

File: subsys/bluetooth/audio/bap_broadcast_assistant.c (hunk 2 above)
Comment (kruithofa) [reply subsys/bluetooth/audio/bap_broadcast_assistant.c:647]: I didn't update the ones in the bap_broadcast_assistant_write_cp_cb, these are called only once. This is similar to what was done for VCP

File: subsys/bluetooth/audio/bap_broadcast_assistant.c [hunk 4]
```diff
@@ -85,6 +85,57 @@ static int16_t lookup_index_by_handle(uint16_t handle)
...
+	SYS_SLIST_FOR_EACH_CONTAINER_SAFE(&broadcast_assistant_cbs,
+					  listener, next, _node) {
+		if (listener->discover) {
+			listener->discover(conn, err, recv_state_count);
+		}
+	}
+}
+
+static void bap_broadcast_assistant_recv_state(struct bt_conn *conn, int err,
+					       const struct bt_bap_scan_delegator_recv_state *s)
```
Comment (Thalley) [inline subsys/bluetooth/audio/bap_broadcast_assistant.c:102]:
```suggestion
static void bap_broadcast_assistant_recv_state_changed(struct bt_conn *conn, int err,
					       const struct bt_bap_scan_delegator_recv_state *s)
```

**Review by Thalley — CHANGES_REQUESTED** [review]
One change request, and a few optional suggestions

File: subsys/bluetooth/audio/bap_broadcast_assistant.c [hunk 5]
```diff
@@ -85,6 +85,57 @@ static int16_t lookup_index_by_handle(uint16_t handle)
...
+{
+	struct bt_bap_broadcast_assistant_cb *listener, *next;
+
+	SYS_SLIST_FOR_EACH_CONTAINER_SAFE(&broadcast_assistant_cbs,
+					  listener, next, _node) {
+		if (listener->recv_state) {
+			listener->recv_state(conn, err, s);
```
[the remark this answers is not in the document]
Comment (kruithofa) [reply subsys/bluetooth/audio/bap_broadcast_assistant.c:109]: I know but there is the 100 char limit

File: subsys/bluetooth/audio/bap_broadcast_assistant.c (hunk 4 above)
Comment (kruithofa) [reply subsys/bluetooth/audio/bap_broadcast_assistant.c:102]: That's what I had first, but then it came over the 100 char limit. So need to choose what is the least bad here...

File: subsys/bluetooth/audio/bap_broadcast_assistant.c [hunk 6]
```diff
@@ -85,6 +85,57 @@ static int16_t lookup_index_by_handle(uint16_t handle)
...
+{
+	struct bt_bap_broadcast_assistant_cb *listener, *next;
+
+	SYS_SLIST_FOR_EACH_CONTAINER_SAFE(&broadcast_assistant_cbs,
+					  listener, next, _node) {
+		if (listener->recv_state_removed) {
+			listener->recv_state_removed(conn, err, src_id);
+		}
+	}
+}
```
[the remark this answers is not in the document]
Comment (kruithofa) [reply subsys/bluetooth/audio/bap_broadcast_assistant.c:125]: That's an API change right? Maybe a separate PR

File: subsys/bluetooth/audio/bap_broadcast_assistant.c [hunk 7]
```diff
@@ -85,6 +85,57 @@ static int16_t lookup_index_by_handle(uint16_t handle)
...
+static void bap_broadcast_assistant_discover_complete(struct bt_conn *conn, int err,
+						      uint8_t recv_state_count)
+{
+	struct bt_bap_broadcast_assistant_cb *listener, *next;
+
+	SYS_SLIST_FOR_EACH_CONTAINER_SAFE(&broadcast_assistant_cbs,
+					  listener, next, _node) {
+		if (listener->discover) {
+			listener->discover(conn, err, recv_state_count);
+		}
```
[the remark this answers is not in the document]
Comment (kruithofa) [reply subsys/bluetooth/audio/bap_broadcast_assistant.c:97]: when `err!=0` the caller already sets `recv_state_count` to 0

File: subsys/bluetooth/audio/bap_broadcast_assistant.c [hunk 8]
```diff
@@ -85,6 +85,58 @@ static int16_t lookup_index_by_handle(uint16_t handle)
...
+	}
+}
+
+static void bap_broadcast_assistant_recv_state_changed(struct bt_conn *conn, int err,
+						       const struct bt_bap_scan_delegator_recv_state
+						       *state)
```
Comment (Thalley) [inline subsys/bluetooth/audio/bap_broadcast_assistant.c:103]:
```suggestion
static void
bap_broadcast_assistant_recv_state_changed(struct bt_conn *conn, int err,
					   const struct bt_bap_scan_delegator_recv_state *state)
```

or
```suggestion
static void bap_broadcast_assistant_recv_state_changed(
	struct bt_conn *conn, int err, const struct bt_bap_scan_delegator_recv_state *state)
```
would be better than adding a newline between the type and variable (consider using clang-format to help format)

**Review by Thalley — CHANGES_REQUESTED** [review]
Minor formatting request

**Review by Thalley — APPROVED** [review]

File: subsys/bluetooth/audio/bap_broadcast_assistant.c (hunk 6 above)
[the remark this answers is not in the document]
Comment (Thalley) [reply subsys/bluetooth/audio/bap_broadcast_assistant.c:125]: https://github.com/zephyrproject-rtos/zephyr/issues/69527

**Review by fredrikdanebjer — APPROVED** [review]

**Pull request merged**